src/messages/layouts/MessageLayout.hpp

```diff
@@ -69,27 +70,30 @@ class MessageLayout : boost::noncopyable
     bool isReplyable() const;
 
 private:
+    // methods
+    void actuallyLayout(int width, MessageElementFlags flags);
+    void updateBuffer(QPixmap *pixmap, int messageIndex, Selection &selection);
+
+    // Create new buffer if required, returning the buffer
+    QPixmap *ensureBuffer(QPainter &painter, int width);
+
     // variables
     MessagePtr message_;
-    std::shared_ptr<MessageLayoutContainer> container_;
-    std::shared_ptr<QPixmap> buffer_{};
+    MessageLayoutContainer container_;
+    std::unique_ptr<QPixmap> buffer_{};
```

Comment:
You might be able to do `QPixmap buffer_;` here and use [`QPixmap::isNull`](https://doc.qt.io/qt-6/qpixmap.html#isNull). This would remove another indirection, because a `QPixmap` is [essentially another shared-pointer](https://github.com/qt/qtbase/blob/7fac7b5b5011a3db9ba778e56f2b34021f6bd148/src/gui/image/qpixmap.h#L126) and constructing a null-pixmap [is cheap](https://github.com/qt/qtbase/blob/7fac7b5b5011a3db9ba778e56f2b34021f6bd148/src/gui/image/qpixmap.cpp#L64-L83). You then need to avoid using the `QPixmap`'s copy-constructor, since this [potentially copies the pixmap-data](https://github.com/qt/qtbase/blob/7fac7b5b5011a3db9ba778e56f2b34021f6bd148/src/gui/image/qpixmap.cpp#L183-L195).

Reply:
follow-up pr, but can be explored if someone wants